vault/logical_system.go

```diff
@@ -635,7 +660,64 @@ type SystemBackend struct {
 	Backend *framework.Backend
 }
 
-// handleCapabilitiesreturns the ACL capabilities of the token for a given path
+// corsStatus returns the current CORS configuration as a logical.Response
+func (b *SystemBackend) corsStatusResponse() (*logical.Response, error) {
+	corsConf := b.Core.corsConfig
+	if corsConf == nil {
+		return nil, errCORSNotConfigured
+	}
+
+	return &logical.Response{
+		Data: map[string]interface{}{
+			"enabled":         corsConf.Enabled,
+			"allowed_origins": strings.Join(corsConf.AllowedOrigins, " "),
+		},
+	}, nil
+}
+
+// handleCORSEnable sets the list of origins that are allowed
+// to make cross-origin requests and sets the CORS enabled flag to true
+func (b *SystemBackend) handleCORSEnable(req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	origins := d.Get("allowed_origins").(string)
+
+	err := b.Core.corsConfig.Enable(origins)
+	if err != nil {
+		return nil, err
+	}
+
+	config := &Config{
+		Name: "cors",
+		Settings: map[string]string{
+			"allowed_origins": origins,
+			"enabled":         "true",
+		},
+	}
+
+	// Update the config
+	if err := b.Core.configStore.SetConfig(config); err != nil {
+		return handleError(err)
+	}
+
+	return b.corsStatusResponse()
+}
+
+// handleCORSDisable clears the allowed origins and sets the CORS enabled flag to false
+func (b *SystemBackend) handleCORSDisable(req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	b.Core.CORSConfig().Disable()
```

Comment:
@naunga The problem is scope; while maybe a `configStore`-style object would be useful, what it should look like and how it should behave is not discussed anywhere here and internally would usually look like a separate RFC, especially to figure out things like whether there should be a cache or not. We'd also want to figure out whether the abstractions are right or even needed vs. using the abstractions that already exist for storing data to the backend (as in, since you need to error check anyways when performing operations here, do you gain much?). Part of the reason that something like this doesn't exist is because there's been no perceived need for it up to this point and I'm not sure that's changed, given that there's only a single entry for CORS that needs to be stored/retrieved, and the same would be the case for most other values I can think of. To me this feels outside the scope of this particular feature PR and like something that should be PR'd separately. And if that were the case, that would need to go in before this (but there's time, it's very unlikely this won't get pushed to 0.7).

@briankassouf I'm interested in your thoughts here around header auditing. Depending on where the headers we care about are being parsed out, we might be storing that info in the audit broker, but we may want to do it earlier. In that case we'd need to lock the config appropriately to get the list of headers to keep...*but* dropping those we don't care about at the HTTP layer may cause undue delay to requests that don't end up being audit logged, which in my mind is an argument for having the audit broker deal with it.